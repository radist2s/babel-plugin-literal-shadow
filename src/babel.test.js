const {stripIndent} = require('common-tags');
const {transformAsync} = require('@babel/core');

const getPlugin = (options) => [
    require.resolve('./babel'),
    options
];

const defaultOptions = {
    root: __dirname,
    filename: __filename,
    babelrc: false,
    configFile: false,
    presets: [
        ['@babel/preset-react', {throwIfNamespace: false, useBuiltIns: true}],
    ],
    plugins: [getPlugin()],
};

const transformCode = (code, options = defaultOptions) =>
    transformAsync(stripIndent(code), options);

const transform = (code) => transformCode(code);

transform.with = (params) => (code) => transformCode(code, params);

describe('babel', () => {
    it('should transform the code with string literal identifier', async () => {
        const {code} = await transform`
      import styled, {css} from 'reshadow';
      import shadowName from '@radist2s/literal-shadow';

      const Foobar = 'div';
      const Bar = 'span';

      const Btn = ({color}) => styled\`
        \${shadowName\`\${Foobar}\`} { background: blue; }

        \${shadowName(Bar)} + \${shadowName(Bar)}[disabled] { fill: red; }

        Foobar { color: \${color}; }
      \`(<Foobar />);

      export default Btn;
      `;

        expect(code).toMatchSnapshot();
    });

    it('should transform the code with Function/Class identifier', async () => {
        const {code} = await transform`
      import styled, {css} from 'reshadow';
      import shadowName from '@radist2s/literal-shadow';

      const Comp = () => <div>Component</div>

      const Btn = ({color}) => styled\`
        \${shadowName\`\${Comp}\`} { background: blue; }

        \${shadowName(Comp)} { fill: red; }

        Comp { color: \${color}; }
      \`(<Foobar />);

      export default Btn;
      `;

        expect(code).toMatchSnapshot();
    });

    it('should transform the code in CSS', async () => {
        const {code} = await transform`
            import {css as cssTemplate} from 'reshadow';
            import shadowName from '@radist2s/literal-shadow';

            const Foo = 'div'

            const styles = cssTemplate\`
            \${shadowName\`\${Foo}\`} { background: blue; }

            \${shadowName(Foo)} { fill: red; }
            \`

            export default styles;
        `;
        expect(code).toMatchSnapshot();
    });

    it('should transform the code from start and at the end', async () => {
        const {code} = await transform`
            import {css as cssTemplate} from 'reshadow';
            import shadowName from '@radist2s/literal-shadow';

            const Foo = 'div'

            const styles1 = cssTemplate\`\${shadowName\`\${Foo}\`} { background: blue; }\`
            const styles2 = cssTemplate\`\${shadowName\`\${Foo}\`} { background: blue; }
            \`
            const styles3 = cssTemplate\`
            \${shadowName\`\${Foo}\`} { background: blue; }\`

            export default {styles1, styles2, styles3};
        `;
        expect(code).toMatchSnapshot();
    });
});
