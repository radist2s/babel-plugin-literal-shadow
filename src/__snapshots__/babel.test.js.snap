// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel should transform the code from start and at the end 1`] = `
"import { css as cssTemplate } from 'reshadow';
const Foo = 'div';
const styles1 = cssTemplate\`Foo { background: blue; }\`;
const styles2 = cssTemplate\`Foo { background: blue; }
\`;
const styles3 = cssTemplate\`
Foo { background: blue; }\`;
export default {
  styles1,
  styles2,
  styles3
};"
`;

exports[`babel should transform the code in CSS 1`] = `
"import { css as cssTemplate } from 'reshadow';
const Foo = 'div';
const styles = cssTemplate\`
Foo { background: blue; }

Foo { fill: red; }
\`;
export default styles;"
`;

exports[`babel should transform the code with Function/Class identifier 1`] = `
"import styled, { css } from 'reshadow';

const Comp = () => /*#__PURE__*/React.createElement(\\"div\\", null, \\"Component\\");

const Btn = ({
  color
}) => styled\`
  Comp { background: blue; }

  Comp { fill: red; }

  Comp { color: \${color}; }
\`( /*#__PURE__*/React.createElement(Foobar, null));

export default Btn;"
`;

exports[`babel should transform the code with string literal identifier 1`] = `
"import styled, { css } from 'reshadow';
const Foobar = 'div';
const Bar = 'span';

const Btn = ({
  color
}) => styled\`
  Foobar { background: blue; }

  Bar + Bar[disabled] { fill: red; }

  Foobar { color: \${color}; }
\`( /*#__PURE__*/React.createElement(Foobar, null));

export default Btn;"
`;

exports[`babel should use options 1`] = `
"import myStyled from 'my-styled-lib';

const Button = () => /*#__PURE__*/React.createElement(Button, null);

const style1 = myStyled\`
    Button {
        color: red;
    }
\`;
const style2 = myStyled\`
    Button { // also possible to use function as Tag inside preparing template
        color: red;
    }
    // This could be used for semantic purposes instead of function call implementation
\`;"
`;
